File: src/main/java/ling572/EmpiricalExpectation.java
package ling572;

import ling572.util.Frequency;

public class EmpiricalExpectation extends Expectation {

	@Override
	protected double getCount(String y, String x) {
		Frequency<String> classCounts = this.rawCounts.get(y);
		return classCounts.get(x);
	}

	@Override
	public void build() {
		for (String y : this.getClasses()) {
			 for (String t : this.getFeaturesInClass(y)) {				 
				 double expectation = this.calcExpectation(y, t);
				 this.setExpectation(y, t, expectation);
			 }
		}
	}

	@Override
	protected double calcExpectation(String y, String t) {
		 Frequency<String> classRawCounts = this.rawCounts.get(y);
		 int rawCount = classRawCounts.get(t);
		 double expectation = (double)rawCount * 1/this.instanceCount;
		 return expectation;
	}
}

File: src/main/scala/ling572/port/EmpiricalExpectation.scala
package ling572.port

class EmpiricalExpectation extends Expectation {

  override protected def getCount(y:String, x:String) = {
    this.rawCounts.get(y) match {
      case Some(count) => count.get(x)
      case None => 0.0
    }
  }

  override def build() {
    for (y <- this.getClasses) {
      for (t <- this.getFeaturesInClass(y)) {
        val expectation = this.calcExpectation(y, t)
        this.setExpectation(y, t, expectation)
      }
    }
  }

  override protected def calcExpectation(y:String, t:String):Double = {
    val rawCount:Double = this.rawCounts.get(y) match {
      case Some(classRawCounts) => classRawCounts.get(t).toDouble
      case None => 0.0
    }
    rawCount * 1/this.instanceCount
  }
}
